src/tls.rs

```diff
@@ -309,13 +371,9 @@ impl MaybeTlsIncoming<Incoming> {
     }
 }
 
-impl<I> Stream for MaybeTlsIncoming<I>
-where
-    I: Stream,
-    I::Item: AsyncRead + AsyncWrite,
-    I::Error: Into<crate::Error>,
-{
-    type Item = MaybeTlsStream<I::Item>;
+#[cfg(feature = "sources-tls")]
+impl Stream for MaybeTlsIncoming<Incoming> {
```

Comment:
Just curious what the reason for making this no longer generic was?

Reply:
Yeah, I found that incredibly messy. In order to do the `fut.poll().context(Handshake)` in the `poll` function, the `Handshake` error type needs to have `source: HandshakeError`. That type is actually a generic, and depends on the type bound of `MaybeTlsIncoming`. Trying to make `TlsError` generic to compensate for this produced problems all over the place with other errors. The simplest fix was to just nail down the `impl Stream` to the actual type we need.

Reply:
Well, that is why I made the source of the error `crate::Error` and then boxed it before wrapping it in the context? That seemed to work for me?

Reply:
Previously, the error type returned by the poll was `native_tls::Error`. With the openssl crate, it is now `HandshakeError<I>`. It is that change that caused the cascade. I'd be happy to leave it generic, but I couldn't figure a way to make it work.

Reply:
No problem, I don't think we need to be generic just yet anyways. Most likely for that error we could have probably just swaped it out for `ErrorStack`? Though I don't know if there are issues with that.